src/libopensc/card-piv.c

```diff
@@ -3064,38 +3226,104 @@ static int piv_match_card_continued(sc_card_t *card)
+			 */
+			case SC_CARD_TYPE_PIV_II_GENERIC: /* i.e. really dont know what this is */
+			case SC_CARD_TYPE_PIV_II_HIST:
+			case SC_CARD_TYPE_PIV_II_GI_DE:
+			case SC_CARD_TYPE_PIV_II_GEMALTO:
+			case SC_CARD_TYPE_PIV_II_OBERTHUR:
+				iccc = piv_process_ccc(card);
+				sc_debug(card->ctx,SC_LOG_DEBUG_MATCH, "PIV_MATCH card->type:%d iccc:%d ccc_flags:%08x CI:%08x r:%d\n",
+						card->type, iccc, priv->ccc_flags, priv->card_issues, r);
+				/* ignore an error? */
+				/* if CCC says it has CAC with PKI on card set to one of the SC_CARD_TYPE_PIV_II_*_DUAL_CAC */
+				if (priv->ccc_flags & PIV_CCC_F3_CAC_PKI) {
+					switch (card->type)  {
+						case SC_CARD_TYPE_PIV_II_GENERIC:
+						case SC_CARD_TYPE_PIV_II_HIST:
+						case SC_CARD_TYPE_PIV_II_GI_DE:
+						    card->type = SC_CARD_TYPE_PIV_II_GI_DE_DUAL_CAC;
+						    priv->card_issues |= CI_DISCOVERY_USELESS;
+						    priv->obj_cache[PIV_OBJ_DISCOVERY].flags |= PIV_OBJ_CACHE_NOT_PRESENT;
+						    break;
+						case SC_CARD_TYPE_PIV_II_GEMALTO_DUAL_CAC:
+						case SC_CARD_TYPE_PIV_II_GEMALTO:
+							card->type = SC_CARD_TYPE_PIV_II_GEMALTO_DUAL_CAC;
+							priv->card_issues |= CI_DISCOVERY_USELESS;
+							priv->obj_cache[PIV_OBJ_DISCOVERY].flags |= PIV_OBJ_CACHE_NOT_PRESENT;
+							break;
+						case SC_CARD_TYPE_PIV_II_OBERTHUR_DUAL_CAC:
+						case SC_CARD_TYPE_PIV_II_OBERTHUR:
+							card->type =  SC_CARD_TYPE_PIV_II_OBERTHUR_DUAL_CAC;
+							priv->card_issues |= CI_DISCOVERY_USELESS;
+							priv->obj_cache[PIV_OBJ_DISCOVERY].flags |= PIV_OBJ_CACHE_NOT_PRESENT;
+							break;
+					}
+				}
+				break;
+
+				/* if user forced it to be one of the CAC types, assume it is CAC */
+			case SC_CARD_TYPE_PIV_II_GI_DE_DUAL_CAC:
+			case SC_CARD_TYPE_PIV_II_GEMALTO_DUAL_CAC:
+			case SC_CARD_TYPE_PIV_II_OBERTHUR_DUAL_CAC:
+				priv->card_issues |= CI_DISCOVERY_USELESS;
+				priv->obj_cache[PIV_OBJ_DISCOVERY].flags |= PIV_OBJ_CACHE_NOT_PRESENT;
+				break;
+			}
+		}
+	sc_debug(card->ctx,SC_LOG_DEBUG_MATCH, "PIV_MATCH card->type:%d i:%d CI:%08x r:%d\n", card->type, i, priv->card_issues, r);
+	if (i >= 0 && (priv->card_issues & CI_DISCOVERY_USELESS) == 0) {
 		/*
-		 * We now know PIV AID is active, test DISCOVERY object 
-		 * Some CAC cards with PIV don't support DISCOVERY and return 
-		 * SC_ERROR_INCORRECT_PARAMETERS. Any error other then 
-		 * SC_ERROR_FILE_NOT_FOUND means we cannot use discovery 
+		 * We now know PIV AID is active, test DISCOVERY object again 
+		 * Some PIV don't support DISCOVERY and return 
+		 * SC_ERROR_INCORRECT_PARAMETERS. Any error 
+		 * including SC_ERROR_FILE_NOT_FOUND means we cannot use discovery 
 		 * to test for active AID.
 		 */
 		int i7e = piv_find_discovery(card);
 
-		if (i7e != 0 && i7e !=  SC_ERROR_FILE_NOT_FOUND) {
+		sc_debug(card->ctx,SC_LOG_DEBUG_MATCH, "PIV_MATCH card->type:%d i7e:%d CI:%08x r:%d\n", card->type, i7e, priv->card_issues, r);
+		if (i7e != 0) {
```

Comment:
`SC_SUCCESS` instead of `0`?

Reply:
Fixed use `if (i7e < 0) {` as is done in other places when testing  return of `piv_find_discovery(card)`